### PR TITLE
Make "test cases" have tags for easier selection of what subset is being considered.

### PR DESCRIPTION
**Pull request by andy31415**
Specifically things like:
  - flaky means a flaky test, generally not run by default
  - in_development for quickly running only tests that we are working on, ignoring all the rest
  - manual for tests that are to be manually executed
  - slow - can filter out tests that sleep too much (like more than 10s was the list I picked)

One can now set arguments like: `--include-tags IN_DEVELOPMENT` to run only in development tests, or `--exclude-tags SLOW` to not run slow tests. Currently only YAML tests flag things, chip-tool tests retain the older "manual/non-manual" logic.

PR now excludes slow tests from running during PR pull requests, however they still run in branch pushes.

### CODE REVIEW

File: scripts/tests/chiptest/__init__.py [hunk 1]
```diff
@@ -116,19 +64,110 @@ def _GetManualTests() -> Set[ManualTest]:
...
-    manualtests.add(ManualTest(yaml="Config_Example.yaml", reason="Example"))
-    manualtests.add(ManualTest(yaml="Config_Variables_Example.yaml", reason="Example"))
-    manualtests.add(ManualTest(yaml="PICS_Example.yaml", reason="Example"))
-    manualtests.add(ManualTest(yaml="Response_Example.yaml", reason="Example"))
-    manualtests.add(ManualTest(yaml="Test_Example.yaml", reason="Example"))
-
-    # Flagged as manual from: src/app/tests/suites/manualTests.json
-    for item in _LoadManualTestsJson(os.path.join(_YAML_TEST_SUITE_PATH, "manualTests.json")):
-        manualtests.add(item)
+    manualtests.add("Config_Example.yaml")
```
Comment (tehampson) [inline scripts/tests/chiptest/__init__.py:67]: Do we want to add example to it's own category. I know something like PICS_Example.yaml will never work, so I wouldn't want that to break someones manual flow in the future

Comment (andy31415) [reply scripts/tests/chiptest/__init__.py:67]: True, I will mark these as completely excluded then.

File: scripts/tests/chiptest/__init__.py [hunk 2]
```diff
@@ -50,6 +50,21 @@ class ManualTest:
...
 
 
+def _IsValidYamlTest(name: str) -> bool:
+    """Check if the given file name is a valid YAML test.
+
+    This returns invalid for examples, simulated and other specific tests.
+    """
+
+    # Simulated tests are not runnable by repl tests, need
+    # separate infrastructure. Exclude theml completely (they are
```
Comment (tehampson) [inline scripts/tests/chiptest/__init__.py:60]: nit, typo `Exclude theml` -> `Exclude them`

**Review by tehampson — APPROVED** [review]

File: scripts/tests/chiptest/__init__.py [hunk 3]
```diff
@@ -39,67 +39,114 @@ class ManualTest:
...
+        "Test_TC_ACL_2_9.yaml",
+        "TestEvents.yaml",
+        "Test_TC_ACE_1_1.yaml",
+        "Test_TC_ACE_1_5.yaml",
+        "Test_TC_SC_5_1.yaml",
+        "Test_TC_SC_5_2.yaml",
+        "TestCommissionerNodeId.yaml",
+        "TestClusterMultiFabric.yaml",
+        "TestGroupMessaging.yaml",
+        "TestMultiAdmin.yaml",
```
Comment (bzbarsky-apple) [inline scripts/tests/chiptest/__init__.py:146]: This lost the documentation about what the reason is that each test is in the list.  Please add those back, as comments?

File: scripts/tests/run_test_suite.py [hunk 4]
```diff
@@ -67,7 +68,13 @@ class RunContext:
     in_unshare: bool
     chip_tool: str
     dry_run: bool
-    manual_handling: ManualHandling
+    runtime: TestRunTime
+
+    # If not empty, include only the specified test tags
+    include_tags: set(TestTag) = field(default_factory={})
+
+    # If not empty, exclude tests tagged with these tagsexclude tests tagged with these tags
```
Comment (bzbarsky-apple) [inline scripts/tests/run_test_suite.py:76]:
```suggestion
    # If not empty, exclude tests tagged with these tags
```

File: scripts/tests/run_test_suite.py [hunk 5]
```diff
@@ -114,11 +121,16 @@ class RunContext:
...
+    '--include-tags',
+    type=click.Choice(TestTag.__members__.keys(), case_sensitive=False),
+    multiple=True,
+    help='What test tags to include when running.',
+)
+@click.option(
+    '--exclude-tags',
+    type=click.Choice(TestTag.__members__.keys(), case_sensitive=False),
+    multiple=True,
+    help='What test tags to exclude when running.',
```
Comment (bzbarsky-apple) [inline scripts/tests/run_test_suite.py:133]: Should probably document whether include or exclude takes priority if some test matches both.

**Review by bzbarsky-apple — APPROVED** [review]

File: scripts/tests/chiptest/__init__.py (hunk 3 above)
Comment (andy31415) [reply scripts/tests/chiptest/__init__.py:146]: Ah ... did not notice notes were added in the mean time. Updated.

**Pull request merged**